### PR TITLE
feat: change ci pipeline feat

### PR DESCRIPTION
**Pull request by gireesh-devtron**
# Description
This PR contains the code changes to enable users to switch ci source types from one to another.

Fixes #

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: internal/sql/repository/pipelineConfig/CiPipelineRepository.go [hunk 1]
```diff
@@ -552,3 +556,15 @@ func (impl CiPipelineRepositoryImpl) FindCiPipelineByAppIdAndEnvIds(appId int, e
 	_, err := impl.dbConnection.Query(&pipelines, query, appId, pg.In(envIds))
 	return pipelines, err
 }
+
+func (impl CiPipelineRepositoryImpl) FindLinkedCiCount(ciPipelineId int) (int, error) {
+	pipeline := &CiPipeline{}
+	cnt, err := impl.dbConnection.Model(pipeline).
+		Where("parent_ci_pipeline = ?", ciPipelineId).
+		Where("deleted = ?", false).
+		Where("ci_pipeline_type != ?", "LINKED_CD").Count()
```
Comment (kripanshdevtron) [inline internal/sql/repository/pipelineConfig/CiPipelineRepository.go:565]: use constant

File: pkg/bean/app.go [hunk 2]
```diff
@@ -267,13 +269,21 @@ type CiMaterialPatchResponse struct {
...
-	IsCloneJob    bool        `json:"isCloneJob,omitempty"`
+	CiPipeline                     *CiPipeline  `json:"ciPipeline"`
+	AppId                          int          `json:"appId,omitempty"`
+	Action                         PatchAction  `json:"action"`
+	AppWorkflowId                  int          `json:"appWorkflowId,omitempty"`
+	UserId                         int32        `json:"-"`
+	IsJob                          bool         `json:"-"`
+	IsCloneJob                     bool         `json:"isCloneJob,omitempty"`
+	SwitchFromCiPipelineId         int          `json:"switchFromCiPipelineId"`
+	SwitchFromExternalCiPipelineId int          `json:"switchFromExternalCiPipelineId"`
```
Comment (kripanshdevtron) [inline pkg/bean/app.go:280]: Why do we need these ??

File: internal/sql/repository/appWorkflow/AppWorkflowRepository.go [hunk 3]
```diff
@@ -76,6 +78,7 @@ const (
 	CIPIPELINE string = "CI_PIPELINE"
 	CDPIPELINE string = "CD_PIPELINE"
 	WEBHOOK    string = "WEBHOOK"
+	LINKED_CD  string = "LINKED_CD"
```
Comment (subhashish-devtron) [inline internal/sql/repository/appWorkflow/AppWorkflowRepository.go:81]: can remove this now

File: pkg/pipeline/BuildPipelineConfigService.go [hunk 4]
```diff
@@ -1264,6 +1272,319 @@ func (impl *CiPipelineConfigServiceImpl) UpdateCiTemplate(updateRequest *bean.Ci
...
+	}
+	return nil
+}
+
+func (impl *CiPipelineConfigServiceImpl) checkForExternalPipelineAndGetID(externalCiPipelineId int, ciPipeline *pipelineConfig.CiPipeline) (string, int) {
+	if externalCiPipelineId != 0 {
+		return appWorkflow.WEBHOOK, externalCiPipelineId
+	} else if ciPipeline.PipelineType == string(bean.LINKED_CD) {
+		return appWorkflow.LINKED_CD, ciPipeline.Id
+	}
```
Comment (subhashish-devtron) [inline pkg/pipeline/BuildPipelineConfigService.go:1425]: not required anymore

File: pkg/pipeline/BuildPipelineConfigService.go [hunk 5]
```diff
@@ -1264,6 +1272,319 @@ func (impl *CiPipelineConfigServiceImpl) UpdateCiTemplate(updateRequest *bean.Ci
...
+	if request.IsSwitchCiPipelineRequest() {
+		err = impl.updateLinkedAppWorkflowMappings(tx, oldAppWorkflowMapping, res.AppWorkflowMapping)
+		if err != nil {
+			impl.logger.Errorw("error in updating app workflow mappings", "err", err)
+			return nil, err
+		}
+		//we don't store ci-pipeline-id in pipeline table for
+		if request.SwitchFromCiPipelineId > 0 {
+			// ciPipeline id is being set in res object in the addpipelineToTemplate method.
+			//
```
Comment (subhashish-devtron) [inline pkg/pipeline/BuildPipelineConfigService.go:1569]: would be cleaner if we move this entire logic to a function and do something like this

```go

	pipelineExists, err := impl.ciPipelineRepository.CheckIfPipelineExistsByNameAndAppId(request.CiPipeline.Name, request.AppId)
	if err != nil && err != pg.ErrNoRows {
		impl.logger.Errorw("error in fetching pipeline by name, FindByName", "err", err, "patch cipipeline name", request.CiPipeline.Name)
		return nil, err
	}

	if pipelineExists {
		impl.logger.Errorw("pipeline name already exist", "err", err, "patch cipipeline name", request.CiPipeline.Name)
		return nil, fmt.Errorf("pipeline name already exist")
	}
	
	if  request.IsSwitchCiPipelineRequest() {
	handleSwitchRequest () // you can again call impl.addpipelineToTemplate(ciConfig) inside
	} else {
res, err := impl.addpipelineToTemplate(ciConfig)
}

File: internal/sql/repository/pipelineConfig/CiPipelineRepository.go (hunk 1 above)
Comment (gireesh-devtron) [reply internal/sql/repository/pipelineConfig/CiPipelineRepository.go:565]: done

File: internal/sql/repository/appWorkflow/AppWorkflowRepository.go (hunk 3 above)
Comment (gireesh-devtron) [reply internal/sql/repository/appWorkflow/AppWorkflowRepository.go:81]: done

File: pkg/pipeline/BuildPipelineConfigService.go (hunk 5 above)
Comment (gireesh-devtron) [reply pkg/pipeline/BuildPipelineConfigService.go]: done

File: pkg/bean/app.go (hunk 2 above)
Comment (gireesh-devtron) [reply pkg/bean/app.go:280]: removed as these are redundant

File: pkg/pipeline/BuildPipelineConfigService.go [hunk 6]
```diff
@@ -1264,6 +1272,321 @@ func (impl *CiPipelineConfigServiceImpl) UpdateCiTemplate(updateRequest *bean.Ci
...
+	//if switchToType == bean.EXTERNAL {
+	//	//external pipeline can only exist with cd-pipeline
+	//	pipeline, err := impl.pipelineRepository.FindByCiPipelineId(switchFromCiPipelineId)
+	//	if err != nil {
+	//		return nil
+	//	}
+	//	if len(pipeline) == 0 {
+	//		return errors.New("this ci-pipeline cannot be converted to external-webhook type as there are no cd-pipelines attached to this ci-pipeline")
+	//	}
+	//}
```
Comment (subhashish-devtron) [inline pkg/pipeline/BuildPipelineConfigService.go:1302]: please check and remove

**Pull request closed without merging**